nengo/builder/optimizer.py

```diff
@@ -0,0 +1,934 @@
+from nengo.utils.stdlib import Timer, WeakKeyDefaultDict, WeakSet
+
+logger = logging.getLogger(__name__)
+
+__all__ = ["optimize"]
+
+
+def optimize(model, dg, max_passes=None):
+    """Optimizes the operator graph by merging operators.
+
+    This reduces the number of iterators to iterate over in slow Python code
+    (as opposed to fast C code). The resulting merged operators will also
+    operate on larger chunks of sequential memory, making better use of CPU
+    caching and prefetching.
+
+    The optimization algorithm has worst case complexity :math:`O(n^2 + e)`,
+    where :math:`n` is the number of operators and :math:`e` is the number
+    of edges in the dependency graph. In practice the run time will be much
+    better because not all :math:`n^2` pairwise combinations of operators
+    will be evaluated. A grouping depending on the operator type and view
+    bases is done with dictionaries. This grouping can be done in amortized
+    linear time and reduces the actual worst-case runtime of the optimization
+    algorithm to :math:`O(gm^2 + e)`, where :math:`g` is the number of groups
+    and :math:`m` is the number of elements in a group. Moreover, information
+    about memory alignment will be used to cut the inner loop short in
+    many cases and gives a runtime much closer to linear in most cases.
+
+    Note that this function modifies both ``model`` and ``dg``.
+
+    Parameters
+    ----------
+    model : `nengo.builder.Model`
+        Builder output to optimize.
+    dg : dict
+        Dict of the form ``{a: {b, c}}`` where ``b`` and ``c`` depend on ``a``,
+        specifying the operator dependency graph of the model.
+    """
+
+    logger.info("Optimizing model...")
+
+    # We try first to merge operators with views only as these have a fixed
+    # order for the memory alignment whereas operators without views could
+    # be merged in a random order. Merging the views of operators will
+    # propagate requirements in the memory ordering via the other
+    # associated signals of the operator to other operators.
+
+    # Once no more operators with views can be merged, we try to merge
+    # operators without views and then try again merging views (because
+    # each operator merge might generate new views).
+
+    single_pass = OpMergePass(dg)
+
+    n_initial_ops = len(dg)
+    cum_duration = 0.
+    before, after = None, None
+    i = 0
+    only_merge_ops_with_view = True
+    while only_merge_ops_with_view or after < before:
+        if max_passes is not None and i >= max_passes:
+            break
+
+        i += 1
+
+        if logger.isEnabledFor(logging.DEBUG):
+            for tp, ops in iteritems(groupby(dg, type)):
+                logger.debug("%s: %d", tp, len(ops))
+
+        only_merge_ops_with_view = before is None or before != after
+        before = len(dg)
+
+        with Timer() as t:
+            single_pass(only_merge_ops_with_view)
+
+        after = len(dg)
```

Comment:
I initially had a similar thought that it would be nice to return an optimized version of the `dg` and `model` rather than mutating them, but as Jan says, it affects performance significantly. There's a difficult tradeoff to make in this PR, as optimized code tends to be less readable than slow but obvious code. I think this PR manages to strike a good balance between readable and efficient, though it's of course good to get more opinions, and is why we want multiple reviewers for each PR 😄